### PR TITLE
fix(provider/aws): Support for specifying lifecycle hooks for AWS accounts

### PR DESCRIPTION
**Pull request by costimuraru**
Fixes https://github.com/spinnaker/spinnaker/issues/4851

### CODE REVIEW

**Comment by costimuraru** [comment]
Validated that this PR is working as designed, using a real aws environment.

Hal command:
```
hal --daemon-endpoint http://spinnaker-spinnaker-halyard:8064 config provider aws account add npe-account --account-id 0000000000
 --assume-role role/spinnakermanaged-infra
 --regions us-east-1,us-west-2

 --launching-lifecycle-hook-default-result CONTINUE
 --launching-lifecycle-hook-heartbeat-timeout-seconds 1200
 --launching-lifecycle-hook-notification-target-arn spinnaker-lifecycle-pending
 --launching-lifecycle-hook-role-arn role/spinnakerasg-infra

 --terminating-lifecycle-hook-default-result CONTINUE
 --terminating-lifecycle-hook-heartbeat-timeout-seconds 7200
 --terminating-lifecycle-hook-notification-target-arn spinnaker-lifecycle-pending
 --terminating-lifecycle-hook-role-arn role/spinnakerasg-infra
```

Screenshot 2019-09-10 19 28 55


Clouddriver logs:
```
2019-09-10 16:27:58.405  INFO 1 --- [pool-3-thread-1] c.n.s.c.data.task.jedis.JedisTask        : [AWS_DEPLOY] Creating lifecycle hook ({LifecycleHookName: testaws-stage-v005-lifecycle-9309f70d-804c-4792-9f03-d46eb6fba34b,AutoScalingGroupName: testaws-stage-v005,LifecycleTransition: autoscaling:EC2_INSTANCE_LAUNCHING,RoleARN: arn:aws:iam::317085423413:role/spinnakerasg-infra,NotificationTargetARN: arn:aws:sns:us-east-1:317085423413:spinnaker-lifecycle-pending,HeartbeatTimeout: 1200,DefaultResult: CONTINUE}) on us-east-1/testaws-stage-v005
2019-09-10 16:27:58.589  INFO 1 --- [pool-3-thread-1] c.n.s.c.data.task.jedis.JedisTask        : [AWS_DEPLOY] Creating lifecycle hook ({LifecycleHookName: testaws-stage-v005-lifecycle-defd805f-a961-406a-89da-da37fcbe61f3,AutoScalingGroupName: testaws-stage-v005,LifecycleTransition: autoscaling:EC2_INSTANCE_TERMINATING,RoleARN: arn:aws:iam::317085423413:role/spinnakerasg-infra,NotificationTargetARN: arn:aws:sns:us-east-1:317085423413:spinnaker-lifecycle-pending,HeartbeatTimeout: 7200,DefaultResult: CONTINUE}) on us-east-1/testaws-stage-v005
```

File: docs/commands.md [hunk 1]
```diff
@@ -6721,11 +6721,23 @@ Example: [http://{{region}}.eureka.url.to.use:8080/eureka-server/v2](http://{{re
 Using {{region}} will make Spinnaker use AWS regions in the hostname to access discovery so that you can have discovery for multiple regions.
  * `--edda`: The endpoint Edda is reachable at. Edda is not a hard dependency of Spinnaker, but is helpful for reducing the request volume against AWS. See [https://github.com/Netflix/edda](https://github.com/Netflix/edda) for more information.
  * `--environment`: The environment name for the account. Many accounts can share the same environment (e.g. dev, test, prod)
+ * `--launching-lifecycle-hook-default-result`: (*Default*: `ABANDON`) Defines the action the Auto Scaling group should take when the lifecycle hook timeout elapses or if an unexpected failure occurs. This parameter can be either CONTINUE or ABANDON. The default value is ABANDON.
+ * `--launching-lifecycle-hook-heartbeat-timeout-seconds`: (*Default*: `3200`) Set the heartbeat timeout for the lifecycle hook. Instances can " +
+          "remain in a wait state for a finite period of time. The default is one hour (3600 seconds).
```
Comment (dotdotdotpaul) [inline docs/commands.md]: The previous line says the default is 3200 seconds, but the documentation says it's 3600.  Needs to be sync'd.

File: docs/commands.md [hunk 2]
```diff
@@ -6774,6 +6786,12 @@ Example: [http://{{region}}.eureka.url.to.use:8080/eureka-server/v2](http://{{re
...
```
Comment (dotdotdotpaul) [inline docs/commands.md]: Same as above -- 3200 vs 3600

File: docs/commands.md [hunk 3]
```diff
@@ -6783,6 +6801,12 @@ Using {{region}} will make Spinnaker use AWS regions in the hostname to access d
  * `--remove-required-group-membership`: Remove this group from the list of required group memberships.
  * `--remove-write-permission`: Remove this permission to from list of write permissions.
  * `--required-group-membership`: A user must be a member of at least one specified group in order to make changes to this account's cloud resources.
+ * `--terminating-lifecycle-hook-default-result`: (*Default*: `ABANDON`) Defines the action the Auto Scaling group should take when the lifecycle hook timeout elapses or if an unexpected failure occurs. This parameter can be either CONTINUE or ABANDON. The default value is ABANDON.
+ * `--terminating-lifecycle-hook-heartbeat-timeout-seconds`: (*Default*: `3200`) Set the heartbeat timeout for the lifecycle hook. Instances can " +
+          "remain in a wait state for a finite period of time. The default is one hour (3600 seconds).
+ * `--terminating-lifecycle-hook-notification-target-arn`: The ARN of the notification target that Amazon EC2 Auto Scaling uses to notify you when an instance is in the transition state for the lifecycle hook. This target can be either an SQS queue or an SNS topic.
```
Comment (dotdotdotpaul) [inline docs/commands.md]: Same 3200 vs 3600

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/aws/AwsAddAccountCommand.java [hunk 4]
```diff
@@ -62,6 +67,46 @@ protected String getProviderName() {
...
 
+  @Parameter(
+      names = "--launching-lifecycle-hook-default-result",
+      description = AwsCommandProperties.HOOK_DEFAULT_VALUE_DESCRIPTION)
+  private String launchingHookDefaultResult = "ABANDON";
+
+  @Parameter(
+      names = "--launching-lifecycle-hook-heartbeat-timeout-seconds",
+      description = AwsCommandProperties.HOOK_HEARTBEAT_TIMEOUT_DESCRIPTION)
+  private Integer launchingHookHeartbeatTimeoutSeconds = 3200;
```
Comment (dotdotdotpaul) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/providers/aws/AwsAddAccountCommand.java]: Here the default is 3200 specifically.  If this should be 3600 (one hour), needs to change.

**Comment by costimuraru** [comment]
Thanks for the feedback, @dotdotdotpaul. I made the changes.

**Comment by costimuraru** [comment]
Performed the actual commit push now :)

**Comment by costimuraru** [comment]
Guys, can we please merge this? Appreciate the help.
ping @dotdotdotpaul @robzienert @clareliguori @imosquera

**Comment by robzienert** [comment]
@costimuraru Looks like the build failed after updating this branch. Once fixed, I'll merge.

**Comment by costimuraru** [comment]
@robzienert done. should be good now

**Pull request merged**